### ISSUE TITLE
No optimization for rerendering when a global state is hooked via nested state

### ISSUE THREAD

**Issue by rsmidt** [avkonst/hookstate#230]
I'm trying to build a "form management" library, that allows to structure the form fields in deeply nested objects like this:
```
{
  "field1": {
    "id": "field1",
    "label": "Label for field1",
    "value": "Initial value for field1"
  },
  "field2": {
    "id": "field2",
    "label": "Label for field2",
    "value": "Initial value for field2"
  },
  "section": {
    "field1": {
      "id": "section.field1",
      "label": "Label for section.field1",
      "value": "Initial value for section.field1"
    }
  }
}
```

The form fields themselves are complex objects, too.

Forms are composed by specifying a `name` attribute on the field components, e.g. `<Field name="deeply.nested.sub.field" />`. Because `state.nested("deeply.nested.sub.field")` is not supported, I wrote a function to split that property path and then iteratively call `nested` on all parts.

When I edit "section.field1" it seems to work, but when I edit "field1" the "section.field1" component also receives an update. You can find the code here: https://codesandbox.io/s/stupefied-joliot-mkkhl?file=/src/App.tsx

**Comment by avkonst** [comment]
Confirmed as potential optimization missing. Here is the reduced reproducer with a workaround (search for '<===' in code):

```
import { createState, useHookstate, State } from "@hookstate/core";
import "./styles.css";

type Field = {
  label: string;
  value: string;
  id: string;
};

type FooMask = {
  field1: Field;
  section: {
    field1: Field;
  };
  field2: Field;
};

const testField = (id: string): Field => ({
  id,
  label: `Label for ${id}`,
  value: `Initial value for ${id}`
});

// Global state of form
const formState = createState<FooMask>({
  field1: testField("field1"),
  field2: testField("field2"),
  section: {
    field1: testField("section.field1")
  }
});

const Label = () => {
  return (
    <div>
      {Date.now()}
    </div>
  );
};

// Convience hoc
// const WithFieldState = (Comp) => ({ name, ...rest }) => {
//   // const state = useHookstate(getNested(formState as any, name));
//   const state = getNested(formState as any, name);
//   return <Comp name={name} fieldState={state} {...rest} />;
// };

type FieldCompProps = {
  name: string;
  fieldState: State<Field>;
};
const FieldComp = ({ fieldState }: FieldCompProps) => {
  const state = useHookstate(fieldState);

  const handleChange = (event) => {
    state.nested("value").set(event.target.value);
  };

  return (
    <div>
      <Label fieldState={state} />
      <input
        type="text"
        value={state.nested("value").get()}
        onChange={handleChange}
      />
    </div>
  );
};

// Compose form of fields, can be dynamic.
const Form = () => {
  let formState1 = formState; // <=== with this optimizes incorrectly!
  // let formState1 = useHookstate(formState); // <=== with this optimizes correctly!
  return (
    <div>
      <div>
        <code>//Edit this: seems fine</code>
      </div>
      <FieldComp fieldState={formState1.section.field1} />
      <div style={{ marginTop: 40 }}>
        <code>//Edit this: seems wrong</code>
      </div>
      <FieldComp fieldState={formState1.field1} />
    </div>
  );
};


export default function App() {
  return (
    <div className="App">
      <Form />
    </div>
  );
}
```

**Comment by rsmidt** [comment]
Thanks for the suggestion. Unfortunately, the `name`-prop pattern is a hard requirement. I will stick with recoil for now.

**Comment by avkonst** [comment]
I didn't not suggest to remove name prop pattern. I just commented it out
to make sure it is not an issue. It was not the cause of the issue. To
workaround: Consume global state in parent form component, and it will
address the issue

**Comment by avkonst** [comment]
your usecase is fully supported with with hookstate 3.0.12 version

**Comment by avkonst** [comment]
verified in your sandbox and in unit tests

**Comment by dbrxnds** [comment]
Hi @rsmidt, I know it's been a while since this issue was initially made, but I was wondering if you ever ended up finishing this form management "library"? 

Been thinking about a similar solution and was curious if you ran into any significant problems, or if you had anything to share.

Thanks!

**Comment by rsmidt** [comment]
Hi @dbrxnds, I unfortunately never came up with a 100% satisfying solution. What I end up doing was:

## 1. Use **jotai** to store the form description.

I've decided to use jotai because I've wanted something more 'stable' than recoil, but that still used the bottom-up state building approach.

I had one atom storing the whole description (see my initial post) and then derived the form field atoms via something like [focus atoms](https://jotai.org/docs/recipes/large-objects#focusatom).

I then had to make the decision if I wanted to also store the `value` inside the atoms and basically build my own form management for dirtying, validating values etc. based on jotai. But because of time constraints, I decided against it.

## 2. Use **react-hook-form** to manage _only_ the `value` of the form field.

react-hook-form is way more bulletproof and also provides subscription-based form field handling, which was required. We used one `Controller` per form field `value`.

## 3. Introduce a component to combine both sources.

Handling those two state sources was quickly getting annoying, so I built a component that (only) took the name property `path.to.my.field` (without `value`).

The component then subscribed to both the jotai atom and the react-hook-form field. This way I had both the field `value` and the static information for the label, server-generated validations/errors etc.

## 4. Handling side effects.

We used a combination of both server-side and client-side validation. The field description included properties like `required` or `regex` that were translated client-side to react-hook-form `rules`.

Server-side stuff was reported by a `messages` field that was only used when the form field was dirtied once (effortless to get that state with react-hook-form).

I hope that clears it up a little bit.

**Comment by dbrxnds** [comment]
Thank you, @rsmidt. I appreciate the write-up. 

Funnily enough I actually attempted something similar as you are describing, also using jotai very recently, but had some issues with it, mostly the stylistic result. Using hook-form for the single values is a clever way of solving a lot of the issues that come with it though, so I may have a look at that at some point. 

https://github.com/dbrxnds/jotai-zod-form

Actually went to look at hookstate to try and avoid the issues I was having with jotai, although those are most likely to do with my lack of knowledge rather than jotai's abilities. 

Either way, still looking for a proper performant type-safe form solution, unfortunately it seems that putting something together yourself still seems to be the best solution.

**Comment by speigg** [comment]
@avkonst  has this optimization been implemented? It seems that it has been implemented by someone:  https://github.com/shinyjohn0401/hookstate/commit/65520ff1770cbf97c7b9f8867fc63f1c80d551cf

We hook into nested global state a lot in our [engine](https://github.com/EtherealEngine/etherealengine), under the assumption that hooking nested state would trigger less re-renders than hooking into the root state (alas, we realized that this optimization only occurs for local state, not global state). If fixing this is as straightforward as the changes in this commit above, it would be fantastic to get this optimization merged into hookstate.
